diff --git a/src/CandidateSet.c b/src/CandidateSet.c
new file mode 100644
index 0000000..3bc9e61
--- /dev/null
+++ b/src/CandidateSet.c
@@ -0,0 +1,65 @@
+// ==================================================================
+// CandidateSet.h
+// Copyright (c) Jonathan Barker, 2002
+// Copyright (c) Martin Larralde, 2025
+// ==================================================================
+// Implementation of type CandidateSet.
+// ==================================================================
+
+#include <stdlib.h>
+
+#include "CandidateSet.h"
+#include "Atom.h"
+#include "Molecule.h"
+#include "Template.h"
+
+// ==================================================================
+// Methods of type CandidateSet
+// ==================================================================
+
+CandidateSet *CandidateSet_create()
+{
+	CandidateSet *S;
+
+	S = (CandidateSet*)malloc(sizeof(CandidateSet));
+    if(!S) return NULL;
+
+	S->coord = NULL;
+	S->atom = NULL;
+	S->count = 0;
+	S->capacity = 0;
+    
+    return S;
+}
+
+CandidateSet *CandidateSet_reuse(CandidateSet *S, int n)
+{
+	S->count = 0;
+
+	if(S->capacity < n)
+	{
+		S->capacity = n;
+		S->atom = realloc(S->atom,sizeof(Atom*)*n);
+		S->coord = realloc(S->coord,sizeof(double*)*n);
+
+		if(!S->atom || !S->coord)
+		{
+			CandidateSet_free(S);
+			return NULL;
+		}
+	}
+
+	return S;
+}
+
+void CandidateSet_free(CandidateSet *S)
+{
+	if(S)
+	{
+		if(S->atom) free(S->atom);
+		if(S->coord) free(S->coord);
+		free(S);
+	}
+}
+
+// ==================================================================
