diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..2977a3b
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,45 @@
+cmake_minimum_required(VERSION 3.10)
+project(Jess)
+
+# --- `alloca` is used in `TessTemplate_candidates` to avoid small char* allocations ---
+include(CheckSymbolExists)
+check_symbol_exists(alloca "alloca.h" HAVE_ALLOCA)
+
+# --- `__thread` local storage is used in `Scanner_reuse` ---
+include(CheckSourceRuns)
+check_source_runs(C
+    "
+    #include <stdlib.h>
+    static __thread int i;
+    int main(int argc, const char** argv) { i = 0; return i; }
+    "
+    HAVE_THREADLOCALSTORAGE
+)
+
+add_library(jess)
+target_sources(jess PRIVATE
+	src/Annulus.c
+	src/Box.c
+	src/Join.c
+	src/Molecule.c
+	src/Super.c
+	src/Atom.c
+	src/KdTree.c
+	src/Region.c
+	src/TessAtom.c
+	src/Jess.c
+	src/Scanner.c
+	src/TessTemplate.c
+	src/CandidateSet.c
+	src/ResIndex.c)
+if(HAVE_ALLOCA)
+	target_compile_definitions(jess PUBLIC -DHAVE_ALLOCA)
+endif()
+if(HAVE_THREADLOCALSTORAGE)
+	target_compile_definitions(jess PUBLIC -DHAVE_THREADLOCALSTORAGE)
+endif()
+
+add_executable(jess_bin)
+target_sources(jess_bin PRIVATE src/Main.c)
+target_link_libraries(jess_bin jess m)
+set_property(TARGET jess_bin PROPERTY OUTPUT_NAME jess)
