diff --git a/src/Template.h b/src/Template.h
index 13f68f1..8cb0ee7 100644
--- a/src/Template.h
+++ b/src/Template.h
@@ -9,14 +9,17 @@
 #define TEMPLATE_H
 
 #include "Atom.h"
+#include "CandidateSet.h"
 
 // ==================================================================
 // Forward declarations
 // ==================================================================
 // Template				The template interface
+// IgnoreChainType		The type of ignore_chain in check
 // ==================================================================
 
 typedef struct _Template Template;
+typedef enum { ignoreNone,ignoreResidues,ignoreAtoms } IgnoreType;
 
 // ==================================================================
 // type Template
@@ -26,6 +29,7 @@ typedef struct _Template Template;
 // match(T,k,A)			True if A matches atom k of T
 // range(T,i,j,a,b)		[*a,*b] <- range of |atom i - atom j|
 // check(T,A,k,ignore_chain)	Check n-ary rules on atom k-1 and 0,...,k-2
+// candidates(T,M,k)	Generate a set of candidate atoms of M matching atom k of T
 // position(T,i)		Position of atom i (example position)
 // name(T)			Returns the symbolic name for the template
 // logE(T,x,n)			Provide an estimate of logE for a hit
@@ -37,11 +41,13 @@ struct _Template
 	int (*count)(const Template*);
 	int (*match)(const Template*,int,const Atom*);
 	int (*range)(const Template*,int,int,double*,double*);
-	int (*check)(const Template*,Atom**,int,int);
+	int (*check)(const Template*,Atom**,int*,int,IgnoreType);
+	void (*candidates)(const Template*, const Molecule*, int, CandidateSet**);
 	const double *(*position)(const Template*,int);
 	const char *(*name)(const Template*);
 	double (*logE)(const Template*,double,int);
 	double (*distWeight)(const Template*,int);
+	Template* (*copy)(const Template*);
 };
 
 // ==================================================================
