diff --git a/src/TessAtom.h b/src/TessAtom.h
index 6480d8b..6377691 100644
--- a/src/TessAtom.h
+++ b/src/TessAtom.h
@@ -17,6 +17,20 @@
 // ==================================================================
 
 typedef struct _TessAtom TessAtom;
+struct _TessAtom
+{
+	int code;
+	int resSeq;
+	int nameCount;
+	int resNameCount;
+	char chainID1; //Riziotis edit
+	char chainID2;
+	//char chainID; 
+	char **name;
+	char **resName;
+	double pos[3];
+	double distWeight;
+};
 
 // ==================================================================
 // Methods of type TessAtom
@@ -39,6 +53,8 @@ extern char TessAtom_chainID1(const TessAtom*);
 extern char TessAtom_chainID2(const TessAtom*);
 extern double TessAtom_distWeight(const TessAtom*);
 //extern char TessAtom_chainID(const TessAtom*);
+extern TessAtom* TessAtom_copy(const TessAtom*);
+extern double TessAtom_distance(const TessAtom*, const TessAtom*);
 
 // ==================================================================
 
