diff --git a/src/TessAtom.h b/src/TessAtom.h
index 6480d8b..7e267eb 100644
--- a/src/TessAtom.h
+++ b/src/TessAtom.h
@@ -39,6 +39,7 @@ extern char TessAtom_chainID1(const TessAtom*);
 extern char TessAtom_chainID2(const TessAtom*);
 extern double TessAtom_distWeight(const TessAtom*);
 //extern char TessAtom_chainID(const TessAtom*);
+extern TessAtom* TessAtom_copy(const TessAtom*);
 
 // ==================================================================
 
